tap_freshdesk/sync.py

```diff
@@ -0,0 +1,88 @@
+
+def write_schemas(stream_id, catalog, selected_streams):
+    """
+    Write the schemas for each stream.
+    """
+    stream_obj = STREAMS[stream_id]()
+
+    if stream_id in selected_streams:
+        # Get catalog object for a particular stream.
+        stream = [cat for cat in catalog['streams'] if cat['tap_stream_id'] == stream_id ][0]
+        singer.write_schema(stream_id, stream['schema'], stream['key_properties'])
+
+    for child in stream_obj.children:
+        write_schemas(child, catalog, selected_streams)
+
+def get_selected_streams(catalog):
+    '''
+    Gets selected streams.  Checks schema's 'selected'
+    first -- and then checks metadata, looking for an empty
+    breadcrumb and mdata with a 'selected' entry
+    '''
+    selected_streams = []
+    for stream in catalog['streams']:
+        stream_metadata = stream['metadata']
+        for entry in stream_metadata:
+            # Stream metadata will have an empty breadcrumb
+            if not entry['breadcrumb'] and entry['metadata'].get('selected',None):
+                selected_streams.append(stream['tap_stream_id'])
+    return selected_streams
+
+def update_currently_syncing(state, stream_name):
+    """
+    Updates currently syncing stream in the state.
+    """
+    if not stream_name and singer.get_currently_syncing(state):
+        del state['currently_syncing']
+    else:
+        singer.set_currently_syncing(state, stream_name)
+    singer.write_state(state)
+
+def get_ordered_stream_list(currently_syncing, streams_to_sync):
+    """
+    Get an ordered list of remaining streams to sync other streams followed by synced streams.
+    """
+    stream_list = list(sorted(streams_to_sync))
+    if currently_syncing in stream_list:
+        index = stream_list.index(currently_syncing)
+        stream_list = stream_list[index:] + stream_list[:index]
+    return stream_list
+
+def get_stream_to_sync(selected_streams):
+    """
+    Get the streams for which the sync function should be called(the parent in case of selected child streams).
+    """
+    streams_to_sync = []
+    for stream_name, stream_obj in STREAMS.items():
+        if (stream_name in selected_streams) or any(child in selected_streams for child in stream_obj.children):
+            streams_to_sync.append(stream_name)
+    return streams_to_sync
+
+def sync(client, config, state, catalog):
+    """
+    Sync selected streams.
+    """
+
+    # Get selected streams, make sure stream dependencies are met
+    selected_streams = get_selected_streams(catalog)
+    streams_to_sync = get_stream_to_sync(selected_streams)
+    LOGGER.info("Selected Streams: %s", selected_streams)
+    LOGGER.info("Syncing Streams: %s", streams_to_sync)
+
+    singer.write_state(state)
+    currently_syncing = singer.get_currently_syncing(state)
+    streams_to_sync = get_ordered_stream_list(currently_syncing, streams_to_sync)
+    for stream in filter(lambda x: STREAMS[x]().parent is None, streams_to_sync):
```

Comment:
No need to use a filter, streams_to_sync should not return child streams, check my comment on the unit test of it.

Reply:
Removed the filter so that the streams_to_sync returns only parent streams